File: examples/06-openmp/example2.c
// =================================================================
//
// File: example2.c
// Author: Pedro Perez
// Description: This file contains the code that searches for the
// 				smallest value stored in an array using OpenMP.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include "utils.h"

#define SIZE 1000000000 //1e9

int min_value(int *array, int size) {
	int i, result;

	result = INT_MAX;
	#pragma omp parallel
	{
		int local_min = result;
		#pragma omp for nowait
		for (i = 0; i < size; i++) {
			local_min = MMIN(local_min, array[i]);
		}

		#pragma omp critical
		{
			result = MMIN(result, local_min);
		}
	}
	return result;
}

int main(int argc, char* argv[]) {
	int i, j, *a, pos, result;
	double ms;

	a = (int *) malloc(sizeof(int) * SIZE);
	random_array(a, SIZE);
	display_array("a", a);

	srand(time(0));
	pos = rand() % SIZE;
	printf("Setting value 0 at %i\n", pos);
	a[pos] = 0;

	printf("Starting...\n");
	ms = 0;
	for (i = 0; i < N; i++) {
		start_timer();

		result = min_value(a, SIZE);

		ms += stop_timer();
	}
	printf("result = %i\n", result);
	printf("avg time = %.5lf ms\n", (ms / N));

	free(a);
	return 0;
}

File: examples/04-fork-join/Example2.java
// =================================================================
//
// File: Example2.java
// Author: Pedro Perez
// Description: This file contains the code that searches for the
// 				smallest value stored in an array using Java's
//				Fork-Join.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================
import java.util.concurrent.RecursiveTask;
import java.util.concurrent.ForkJoinPool;

public class Example2 extends RecursiveTask<Integer> {
	private static final int SIZE = 100_000_000;
	private static final int MIN = 10_000;
	private int array[], start, end;

	public Example2(int array[], int start, int end) {
		this.array = array;
		this.start = start;
		this.end = end;
	}

	public Integer computeDirectly() {
		int result = Integer.MAX_VALUE;
		for (int i = start; i < end; i++) {
			result = (int) Math.min(result, array[i]);
		}
		return result;
	}

	@Override
	protected Integer compute() {
		if ( (end - start) <= MIN ) {
			return computeDirectly();
		} else {
			int mid = start + ( (end - start) / 2 );
			Example2 lowerMid = new Example2(array, start, mid);
			lowerMid.fork();
			Example2 upperMid = new Example2(array, mid, end);
			return ((int) Math.min(upperMid.compute(), lowerMid.join()));
		}
	}

	public static void main(String args[]) {
		long startTime, stopTime;
		int array[], result = 0;
		double ms;
		ForkJoinPool pool;

		array = new int[SIZE];
		Utils.randomArray(array);
		Utils.displayArray("array", array);

		int pos = Math.abs(Utils.r.nextInt()) % SIZE;
		System.out.printf("Setting value 0 at %d\n", pos);
		array[pos] = 0;

		System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
		ms = 0;
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();

			pool = new ForkJoinPool(Utils.MAXTHREADS);
			result = pool.invoke(new Example2(array, 0, array.length));

			stopTime = System.currentTimeMillis();
			ms += (stopTime - startTime);
		}
		System.out.printf("result = %d\n", result);
		System.out.printf("avg time = %.5f\n", (ms / Utils.N));
	}
}
